File: bindings/android/src/commonMain/kotlin/ai/uldren/loom/facets/contacts.kt
package ai.uldren.loom

    /**
     * Create (or replace the metadata of) address book [book] under [principal] in [workspace] (UUID or
     * name, created with the `contacts` facet if absent).
     */
expect fun Loom.cardCreateBook(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        displayName: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Delete address book [book] under [principal] and its contacts; returns whether it existed. */
expect fun Loom.cardDeleteBook(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): Boolean


    /** List the address-book ids under [principal] as Loom Canonical CBOR (array of text strings). */
expect fun Loom.cardListBooks(
        path: String,
        workspace: String,
        principal: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** Put the contact [entry] (its `ContactEntry` canonical CBOR) into [book], keyed by its UID. */
expect fun Loom.cardPutEntry(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        entry: ByteArray,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Fetch the contact at [uid] in [book] as its `ContactEntry` canonical CBOR, or null if absent. */
expect fun Loom.cardGetEntry(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        uid: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray?


    /** Remove the contact at [uid] in [book]; returns whether it was present. */
expect fun Loom.cardDeleteEntry(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        uid: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): Boolean


    /** List [book] as Loom Canonical CBOR (array of per-contact `ContactEntry` CBOR byte strings). */
expect fun Loom.cardListEntries(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /**
     * Search [book] by a case-insensitive substring [text] over name, organization, and email; returns
     * Loom Canonical CBOR (array of per-contact `ContactEntry` CBOR byte strings).
     */
expect fun Loom.cardSearch(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        text: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** The on-demand vCard (`.vcf`) projection of the contact at [uid], or null if absent. */
expect fun Loom.cardEntryVcard(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        uid: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): String?


    /** Parse vCard [vcf] and store it as a record in [book]; returns the new ETag ("algo:hex"). */
expect fun Loom.cardPutVcard(
        path: String,
        workspace: String,
        principal: String,
        book: String,
        vcf: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): String

File: bindings/react-native/src/facets/contacts.ts
import UldrenLoom from '../NativeUldrenLoom';
import { authArgs, keyArgs } from '../internal';
import type { LoomAuth, LoomKey } from '../internal';

/**
 * Create (or replace the metadata of) address book `book` under `principal` in `workspace` (UUID or name,
 * created with the `contacts` facet if absent).
 */
export function cardCreateBook(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  displayName: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.cardCreateBook(loomPath, workspace, principal, book, displayName, passphrase, kek, authPrincipal, authPassphrase);
}

/** Delete address book `book` under `principal` and its contacts; resolves whether it existed. */
export function cardDeleteBook(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<boolean> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.cardDeleteBook(loomPath, workspace, principal, book, passphrase, kek, authPrincipal, authPassphrase);
}

/** The address-book ids under `principal` as raw Loom Canonical CBOR (an array of text strings). */
export async function cardListBooksCbor(
  loomPath: string,
  workspace: string,
  principal: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.cardListBooks(loomPath, workspace, principal, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/** Put the contact `entry` (its `ContactEntry` canonical CBOR) into `book`, keyed by its UID. */
export function cardPutEntry(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  entry: Uint8Array | number[],
  key?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.cardPutEntry(loomPath, workspace, principal, book, Array.from(entry), passphrase, kek, authPrincipal, authPassphrase);
}

/** Fetch the contact at `uid` in `book` as its `ContactEntry` canonical CBOR, or null if absent. */
export async function cardGetEntry(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  uid: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array | null> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.cardGetEntry(loomPath, workspace, principal, book, uid, passphrase, kek, authPrincipal, authPassphrase);
  return bytes == null ? null : Uint8Array.from(bytes);
}

/** Remove the contact at `uid` in `book`; resolves whether it was present. */
export function cardDeleteEntry(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  uid: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<boolean> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.cardDeleteEntry(loomPath, workspace, principal, book, uid, passphrase, kek, authPrincipal, authPassphrase);
}

/** List `book` as raw Loom Canonical CBOR (an array of per-contact `ContactEntry` CBOR byte strings). */
export async function cardListEntriesCbor(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.cardListEntries(loomPath, workspace, principal, book, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/**
 * Search `book` by a case-insensitive substring `text` over name, organization, and email; resolves raw
 * Loom Canonical CBOR (an array of per-contact `ContactEntry` CBOR byte strings).
 */
export async function cardSearchCbor(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  text: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.cardSearch(loomPath, workspace, principal, book, text, passphrase, kek, authPrincipal, authPassphrase);
  return Uint8Array.from(bytes);
}

/** The on-demand vCard (`.vcf`) projection of the contact at `uid`, or null if absent. */
export function cardEntryVcard(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  uid: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<string | null> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.cardEntryVcard(loomPath, workspace, principal, book, uid, passphrase, kek, authPrincipal, authPassphrase);
}

/** Parse vCard `vcf` and store it as a record in `book`; resolves the new ETag ("algo:hex"). */
export function cardPutVcard(
  loomPath: string,
  workspace: string,
  principal: string,
  book: string,
  vcf: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<string> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.cardPutVcard(loomPath, workspace, principal, book, vcf, passphrase, kek, authPrincipal, authPassphrase);
}
